File: app/src/main/java/com/kelsos/mbrc/utilities/SocketActivityChecker.java
package com.kelsos.mbrc.utilities;

import android.support.annotation.NonNull;
import com.google.inject.Singleton;
import java.util.concurrent.TimeUnit;
import rx.Completable;
import rx.Subscription;
import timber.log.Timber;

@Singleton
public class SocketActivityChecker {
  private static final int DELAY = 40;
  private Subscription subscription;
  private PingTimeoutListener pingTimeoutListener;

  public SocketActivityChecker() {

  }

  public void start() {
    Timber.v("Starting activity checker");
    subscription = getSubscribe();
  }

  @NonNull
  private Subscription getSubscribe() {
    return Completable.timer(DELAY, TimeUnit.SECONDS).subscribe(throwable -> Timber.v("Subscription failed"), () -> {
      Timber.v("Ping was more than %d seconds ago", DELAY);
      if (pingTimeoutListener == null) {
        return;
      }
      pingTimeoutListener.onTimeout();
    });
  }

  public void stop() {
    Timber.v("Stopping activity checker");
    unsubscribe();
  }

  public void ping() {
    Timber.v("Received ping");
    unsubscribe();
    subscription = getSubscribe();
  }

  private void unsubscribe() {
    if (subscription != null && !subscription.isUnsubscribed()) {
      subscription.unsubscribe();
    }
  }

  public void setPingTimeoutListener(PingTimeoutListener pingTimeoutListener) {
    this.pingTimeoutListener = pingTimeoutListener;
  }

  public interface PingTimeoutListener {
    void onTimeout();
  }
}

File: app/src/main/java/com/kelsos/mbrc/utilities/SocketActivityChecker.kt
package com.kelsos.mbrc.utilities

import rx.Completable
import rx.Subscription
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

@Singleton
class SocketActivityChecker {
  private var subscription: Subscription? = null
  private var pingTimeoutListener: PingTimeoutListener? = null

  fun start() {
    Timber.v("Starting activity checker")
    subscription = subscribe
  }

  private val subscribe: Subscription
    get() = Completable.timer(DELAY.toLong(), TimeUnit.SECONDS).subscribe({
      Timber.v("Ping was more than %d seconds ago", DELAY)
      pingTimeoutListener?.onTimeout()
    }) { Timber.v("Subscription failed") }

  fun stop() {
    Timber.v("Stopping activity checker")
    unsubscribe()
  }

  fun ping() {
    Timber.v("Received ping")
    unsubscribe()
    subscription = subscribe
  }

  private fun unsubscribe() {
    if (subscription != null && !subscription!!.isUnsubscribed) {
      subscription!!.unsubscribe()
    }
  }

  fun setPingTimeoutListener(pingTimeoutListener: PingTimeoutListener?) {
    this.pingTimeoutListener = pingTimeoutListener
  }

  interface PingTimeoutListener {
    fun onTimeout()
  }

  companion object {
    private const val DELAY = 40
  }
}
